src/dev.rs

```diff
@@ -0,0 +1,343 @@
+                        self.aux[column.index()][resolved_row as usize].clone()
+                    },
+                    &|a, b| a + &b,
+                    &|a, b| a * &b,
+                    &|a, scalar| a * scalar,
+                ) != F::zero()
+                {
+                    return Err(VerifyFailure::Gate {
+                        gate_index,
+                        row: (row - n) as usize,
+                    });
+                }
+            }
+        }
+
+        // Check that all lookups exist in their respective tables.
+        for (lookup_index, lookup) in self.cs.lookups.iter().enumerate() {
+            for input_row in 0..n {
+                let load = |column: &Column<Any>, row| match column.column_type() {
+                    Any::Fixed => self.fixed[column.index()][row as usize],
+                    Any::Advice => self.advice[column.index()][row as usize],
+                    Any::Aux => self.aux[column.index()][row as usize],
+                };
+
+                let inputs: Vec<_> = lookup
+                    .input_columns
+                    .iter()
+                    .map(|c| load(c, input_row))
+                    .collect();
+                if !(0..n)
+                    .map(|table_row| lookup.table_columns.iter().map(move |c| load(c, table_row)))
+                    .any(|table_row| table_row.eq(inputs.iter().cloned()))
+                {
+                    return Err(VerifyFailure::Lookup {
+                        lookup_index,
+                        row: input_row as usize,
+                    });
+                }
+            }
+        }
+
+        // Check that permutations preserve the original values of the cells.
+        for (perm_index, assembly) in self.permutations.iter().enumerate() {
+            // Original values of columns involved in the permutation
+            let original = self.cs.permutations[perm_index]
+                .get_columns()
+                .iter()
+                .map(|c| self.advice[c.index()].clone())
+                .collect::<Vec<_>>();
+
+            // Iterate over each column of the permutation
+            for (column, values) in assembly.mapping.iter().enumerate() {
+                // Iterate over each row of the column to check that the cell's
+                // value is preserved by the mapping.
+                for (row, cell) in values.iter().enumerate() {
+                    let original_cell = original[column][row];
+                    let permuted_cell = original[cell.0][cell.1];
+                    if original_cell != permuted_cell {
+                        return Err(VerifyFailure::Permutation {
+                            perm_index,
+                            column,
+                            row,
+                        });
+                    }
+                }
+            }
+        }
+
+        // TODO: Implement the rest of the verification checks.
```

Comment:
There shouldn't be any additional checks right?

Reply:
Now that the permutation checks have been added, I think this comment is moot.